File: templates/problems/04_stack/01_valid_parentheses/solution.go
package valid_parentheses

// solveIsValid checks if brackets are valid using a stack. O(n) time, O(n) space.
func solveIsValid(s string) bool {
	stack := []byte{}
	pairs := map[byte]byte{')': '(', ']': '[', '}': '{'}
	for i := 0; i < len(s); i++ {
		c := s[i]
		if c == '(' || c == '[' || c == '{' {
			stack = append(stack, c)
		} else {
			if len(stack) == 0 || stack[len(stack)-1] != pairs[c] {
				return false
			}
			stack = stack[:len(stack)-1]
		}
	}
	return len(stack) == 0
}

File: templates/problems/04_stack/01_valid_parentheses/typescript/solution.ts
/**
 * Check if brackets are valid using a stack.
 * Time: O(n), Space: O(n)
 */
export function solveIsValid(s: string): boolean {
  const stack: string[] = [];
  const pairs: Record<string, string> = { ')': '(', ']': '[', '}': '{' };
  for (const c of s) {
    if (c === '(' || c === '[' || c === '{') {
      stack.push(c);
    } else {
      if (stack.length === 0 || stack[stack.length - 1] !== pairs[c]) {
        return false;
      }
      stack.pop();
    }
  }
  return stack.length === 0;
}
